### ISSUE TITLE
UAA creates pid file before migrations complete and server is ready

### ISSUE THREAD

**Issue by sykesm** [cloudfoundry/uaa#150]
The monit template for UAA uses a [pidfile][pidfile] as the primary process check and the UAA control script creates that file [before][before] it execs java. This combination can be very problematic as bosh will determine the UAA is alive before it is ready to accept requests; this is especially true when UAADB migrations are pending.

In particular, if there's a long running database migration, bosh will detect that the UAA canary is alive and well as soon as monit says it's active (within a few seconds) - even if migrations are still running and the server is not accepting requests.  This is due to the [10 minutes][failed] of grace that the monit config declares.

If you look at the monit log during a long migration, you will see the following:
```
[UTC May 19 01:00:35] info     : 'system_b9362d58-aa27-4a4d-851f-c2ffcd6f6790' Monit reloaded
[UTC May 19 01:00:36] error    : 'uaa' process is not running
[UTC May 19 01:00:36] info     : 'uaa' trying to restart
[UTC May 19 01:00:36] info     : 'uaa' start: /var/vcap/jobs/uaa/bin/uaa_ctl
[UTC May 19 01:00:37] info     : 'uaa_cf-registrar' start: /var/vcap/jobs/uaa/bin/uaa_cf-registrar_ctl
[UTC May 19 01:00:38] error    : 'metron_agent' process is not running
[UTC May 19 01:00:38] info     : 'metron_agent' trying to restart
[UTC May 19 01:00:38] info     : 'metron_agent' start: /var/vcap/jobs/metron_agent/bin/metron_agent_ctl
[UTC May 19 01:00:39] error    : 'consul_agent' process is not running
[UTC May 19 01:00:39] info     : 'consul_agent' trying to restart
[UTC May 19 01:00:39] info     : 'consul_agent' start: /var/vcap/jobs/consul_agent/bin/agent_ctl
[UTC May 19 01:00:40] info     : start service 'consul_agent' on user request
[UTC May 19 01:00:40] info     : monit daemon at 77 awakened
[UTC May 19 01:00:40] info     : Awakened by User defined signal 1
[UTC May 19 01:00:40] info     : 'consul_agent' start action done
[UTC May 19 01:00:40] info     : start service 'metron_agent' on user request
[UTC May 19 01:00:40] info     : monit daemon at 77 awakened
[UTC May 19 01:00:40] info     : start service 'uaa_cf-registrar' on user request
[UTC May 19 01:00:40] info     : monit daemon at 77 awakened
[UTC May 19 01:00:40] info     : start service 'uaa' on user request
[UTC May 19 01:00:40] info     : monit daemon at 77 awakened
[UTC May 19 01:00:41] info     : 'uaa' start action done
[UTC May 19 01:00:41] info     : 'uaa_cf-registrar' start action done
[UTC May 19 01:00:41] info     : 'metron_agent' start action done
[UTC May 19 01:00:41] info     : 'consul_agent' process is running with pid 476
[UTC May 19 01:00:41] info     : Awakened by User defined signal 1
[UTC May 19 01:00:41] info     : 'uaa' process is running with pid 142
[UTC May 19 01:01:41] error    : HTTP: error receiving data -- Resource temporarily unavailable
[UTC May 19 01:01:41] error    : 'uaa' failed protocol test [HTTP] at INET[localhost:8080/healthz] via TCP
[UTC May 19 01:01:41] info     : 'metron_agent' process is running with pid 426
[UTC May 19 01:02:51] error    : HTTP: error receiving data -- Resource temporarily unavailable
[UTC May 19 01:02:51] error    : 'uaa' failed protocol test [HTTP] at INET[localhost:8080/healthz] via TCP
[UTC May 19 01:04:01] error    : HTTP: error receiving data -- Resource temporarily unavailable
[UTC May 19 01:04:01] error    : 'uaa' failed protocol test [HTTP] at INET[localhost:8080/healthz] via TCP
[UTC May 19 01:05:11] error    : HTTP: error receiving data -- Resource temporarily unavailable
[UTC May 19 01:05:11] error    : 'uaa' failed protocol test [HTTP] at INET[localhost:8080/healthz] via TCP
[UTC May 19 01:06:21] error    : HTTP: error receiving data -- Resource temporarily unavailable
[UTC May 19 01:06:21] error    : 'uaa' failed protocol test [HTTP] at INET[localhost:8080/healthz] via TCP
[UTC May 19 01:07:31] error    : HTTP: error receiving data -- Resource temporarily unavailable
[UTC May 19 01:07:31] error    : 'uaa' failed protocol test [HTTP] at INET[localhost:8080/healthz] via TCP
[UTC May 19 01:08:41] error    : HTTP: error receiving data -- Resource temporarily unavailable
[UTC May 19 01:08:41] error    : 'uaa' failed protocol test [HTTP] at INET[localhost:8080/healthz] via TCP
[UTC May 19 01:09:51] error    : HTTP: error receiving data -- Resource temporarily unavailable
[UTC May 19 01:09:51] error    : 'uaa' failed protocol test [HTTP] at INET[localhost:8080/healthz] via TCP
[UTC May 19 01:11:01] error    : HTTP: error receiving data -- Resource temporarily unavailable
[UTC May 19 01:11:01] error    : 'uaa' failed protocol test [HTTP] at INET[localhost:8080/healthz] via TCP
[UTC May 19 01:12:11] error    : HTTP: error receiving data -- Resource temporarily unavailable
[UTC May 19 01:12:11] error    : 'uaa' failed protocol test [HTTP] at INET[localhost:8080/healthz] via TCP
[UTC May 19 01:12:11] info     : 'uaa' trying to restart
[UTC May 19 01:12:11] info     : 'uaa_cf-registrar' stop: /var/vcap/jobs/uaa/bin/uaa_cf-registrar_ctl
[UTC May 19 01:12:12] info     : 'uaa' stop: /var/vcap/jobs/uaa/bin/uaa_ctl
[UTC May 19 01:12:13] info     : 'uaa' start: /var/vcap/jobs/uaa/bin/uaa_ctl
```

After the initial 10 minutes have passed, you will see the UAA repeatedly try to restart (uaa_ctl.err.log):
```
[2015-05-19 01:00:36+0000] ------------ STARTING uaa_ctl at Tue May 19 01:00:36 UTC 2015 --------------
[2015-05-19 01:12:13+0000] ------------ STARTING uaa_ctl at Tue May 19 01:12:13 UTC 2015 --------------
[2015-05-19 01:13:27+0000] ------------ STARTING uaa_ctl at Tue May 19 01:13:27 UTC 2015 --------------
[2015-05-19 01:14:41+0000] ------------ STARTING uaa_ctl at Tue May 19 01:14:41 UTC 2015 --------------
[2015-05-19 01:15:56+0000] ------------ STARTING uaa_ctl at Tue May 19 01:15:56 UTC 2015 --------------
[2015-05-19 01:17:10+0000] ------------ STARTING uaa_ctl at Tue May 19 01:17:10 UTC 2015 --------------
[2015-05-19 01:18:24+0000] ------------ STARTING uaa_ctl at Tue May 19 01:18:24 UTC 2015 --------------
[2015-05-19 01:19:38+0000] ------------ STARTING uaa_ctl at Tue May 19 01:19:38 UTC 2015 --------------
[2015-05-19 01:20:52+0000] ------------ STARTING uaa_ctl at Tue May 19 01:20:52 UTC 2015 --------------
[2015-05-19 01:22:06+0000] ------------ STARTING uaa_ctl at Tue May 19 01:22:06 UTC 2015 --------------
[2015-05-19 01:23:20+0000] ------------ STARTING uaa_ctl at Tue May 19 01:23:20 UTC 2015 --------------
[2015-05-19 01:24:34+0000] ------------ STARTING uaa_ctl at Tue May 19 01:24:34 UTC 2015 --------------
[2015-05-19 01:25:48+0000] ------------ STARTING uaa_ctl at Tue May 19 01:25:48 UTC 2015 --------------
[2015-05-19 01:27:02+0000] ------------ STARTING uaa_ctl at Tue May 19 01:27:02 UTC 2015 --------------
[2015-05-19 01:28:16+0000] ------------ STARTING uaa_ctl at Tue May 19 01:28:16 UTC 2015 --------------
[2015-05-19 01:29:30+0000] ------------ STARTING uaa_ctl at Tue May 19 01:29:30 UTC 2015 --------------
[2015-05-19 01:30:44+0000] ------------ STARTING uaa_ctl at Tue May 19 01:30:44 UTC 2015 --------------
[2015-05-19 01:31:58+0000] ------------ STARTING uaa_ctl at Tue May 19 01:31:58 UTC 2015 --------------
[2015-05-19 01:33:12+0000] ------------ STARTING uaa_ctl at Tue May 19 01:33:12 UTC 2015 --------------
[2015-05-19 01:34:26+0000] ------------ STARTING uaa_ctl at Tue May 19 01:34:26 UTC 2015 --------------
[2015-05-19 01:35:40+0000] ------------ STARTING uaa_ctl at Tue May 19 01:35:40 UTC 2015 --------------
[2015-05-19 01:36:55+0000] ------------ STARTING uaa_ctl at Tue May 19 01:36:55 UTC 2015 --------------
```

This, in turn, ensures migration never completes and causes multiple migration statements to run. For example, on Postgres, listing the current queries after a few iterations yields the following (related to the 1.8.4 migration):
```
uaadb=> SELECT datname,procpid,current_query FROM pg_stat_activity;
 datname | procpid |                                   current_query                                    
---------+---------+------------------------------------------------------------------------------------
 ccdb    |     731 | <insufficient privilege>
 ccdb    |     733 | <insufficient privilege>
 ccdb    |     734 | <insufficient privilege>
 ccdb    |     736 | <insufficient privilege>
 ccdb    |     737 | <insufficient privilege>
 uaadb   |    1097 | <IDLE>
 uaadb   |    1098 | <IDLE>
 uaadb   |    1099 | <IDLE>
 uaadb   |    1100 | <IDLE>
 uaadb   |    1101 | <IDLE>
 uaadb   |    1102 | <IDLE>
 uaadb   |    1103 | <IDLE>
 uaadb   |    1104 | <IDLE>
 uaadb   |     979 | ALTER TABLE oauth_client_details ADD COLUMN autoapprove VARCHAR(1024) default NULL
 uaadb   |     993 | ALTER TABLE oauth_client_details ADD COLUMN autoapprove VARCHAR(1024) default NULL
 uaadb   |    1006 | ALTER TABLE oauth_client_details ADD COLUMN autoapprove VARCHAR(1024) default NULL
 uaadb   |    1020 | ALTER TABLE oauth_client_details ADD COLUMN autoapprove VARCHAR(1024) default NULL
 uaadb   |    1034 | ALTER TABLE oauth_client_details ADD COLUMN autoapprove VARCHAR(1024) default NULL
 uaadb   |    1047 | ALTER TABLE oauth_client_details ADD COLUMN autoapprove VARCHAR(1024) default NULL
 uaadb   |    1061 | ALTER TABLE oauth_client_details ADD COLUMN autoapprove VARCHAR(1024) default NULL
 uaadb   |    1075 | ALTER TABLE oauth_client_details ADD COLUMN autoapprove VARCHAR(1024) default NULL
 uaadb   |    1090 | ALTER TABLE oauth_client_details ADD COLUMN autoapprove VARCHAR(1024) default NULL
 uaadb   |    1095 | SELECT datname,procpid,current_query FROM pg_stat_activity;
 uaadb   |    1105 | ALTER TABLE oauth_client_details ADD COLUMN autoapprove VARCHAR(1024) default NULL
 uaadb   |    1106 | <IDLE> in transaction
```

Even when running with multiple UAA instances, since the canary instance was successful, the other instances will eventually start the migrations and fail in the same way. As you can see, this failure mode during migrations can lead to extended outages.

Recreating this issue with Postgres is trivial. Simply create a migration that includes the following statement: `SELECT pg_sleep(900);` That will simulate a very long running migration that exceeds the 10 minute window in monit.

If the pid file creation were deferred until migrations completed, that would have prevented this scenario as it would have resulted in a canary failure.  This would be similar to how cloud controller migrations are run. (To be fair, there are problems there as well.)

[pidfile]: https://github.com/cloudfoundry/cf-release/blob/89da1873f23c17d60e177f953015fe0cfaf95f7d/jobs/uaa/monit#L2
[before]: https://github.com/cloudfoundry/cf-release/blob/89da1873f23c17d60e177f953015fe0cfaf95f7d/jobs/uaa/templates/uaa_ctl.erb#L66-L67
[failed]: https://github.com/cloudfoundry/cf-release/blob/89da1873f23c17d60e177f953015fe0cfaf95f7d/jobs/uaa/monit#L6-L9

**Comment by fhanik** [comment]
I agree, we are actually going to be switching to leveraging Tomcat's PID file utility.
This is done by setting CATALINA_PID variable to a file path. And when Tomcat shuts down, it can force a kill (we've had instances where the old tomcat instance never shut down)

We need to investigate how we can avoid monit to create the file for us.

**Comment by sykesm** [comment]
@fhanik, monit doesn't actually create the file - it simply monitors it. The file is currently created by the [control script](https://github.com/cloudfoundry/cf-release/blob/89da1873f23c17d60e177f953015fe0cfaf95f7d/jobs/uaa/templates/uaa_ctl.erb#L66-L67) in cf-release.

Not sure if that helps or not...

**Comment by fhanik** [comment]
Sure does, we'll get to it shortly.

**Comment by sykesm** [comment]
@sreetummidi: I noticed this was prioritized, picked up, and then iceboxed again. Was there a technical issue that caused the story to be put down? Is there something we can do to help this along?

**Comment by fhanik** [comment]
@sykesm 
We found a lot of wholes around the entire monit system that it got ice boxed because we need to address the entire spectrum of issues.

We currently have the following issues that we want to address, (and we're not sure that monit or cf-release can fully do that yet as it looks today)
  - Ability to run DB migrations before we start up the web server (some migrations can be long running) and by doing this not confusing a migration with an application startup time. Two different timeouts.
  - Ability to not register the route with the router until the web application has completely started up
  - We MUST have the ability to unregister a route when we plan to stop the web application. Without this, we have no way of doing a graceful shutdown during upgrades

What we are seeing today is that some steps of bosh/monit are executed in non blocking manner, thus we have no real contract between different tasks.

So we have decided that before we dig into this, let's hash out the requirements and implement a complete solution rather than small stabs that still don't meet our requirements.

**Comment by sykesm** [comment]
@fhanik, it would have been spectacular if that update were added to the story so interested parties could follow along.

What forums are being used to work out the details? How do we engage to help get this resolved?

Thanks.

**Comment by fhanik** [comment]
@sykesm Our bad for not putting a comment in the story when moving it back to icebox. @sreetummidi we can be a bit more diligent about that.

Currently there are three or four stories that are floating around. I am going to make the recommendation that we may open a completely new one, and close all of the minor ones and simply point it to the new one. That way, anyone can follow along for everything that is going on. 

We're still in the investigation portion of this, yesterday it was brought up that deregistering a route API may have been removed or not working. We need confirmation for that as well before we `start` considering solutions to the story.

**Comment by sreetummidi** [comment]
Apologize for the confusion caused. 
We are working on a doc outlining the current issues with the scenarios below and mitigation paths. We will be able to share this with the community momentarily.
Agree that we need to create a new story and close the old ones appropriately.

**Comment by sreetummidi** [comment]
Fixed starting with : https://github.com/cloudfoundry/uaa-release/releases/tag/v7
